### PR TITLE
ForegroundDispatcher improvements

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ForegroundDispatcher.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.Razor
+{
+    internal abstract class ForegroundDispatcher : IWorkspaceService
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ForegroundDispatcher.cs]: Moved from language services and added TaskSchedulers

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundFactAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Test
+{
+    // Similar to WpfFactAttribute https://github.com/xunit/samples.xunit/blob/969d9f7e887836f01a6c525324bf3db55658c28f/STAExamples/WpfFactAttribute.cs
```
Comment (NTaylorMullen) [inline test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundFactAttribute.cs]: Dead link. Also, I'm sure we have some test that ensure foreground. Would be nice to have an example.

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundDispatcherTestBase.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        private class ThrowingTaskScheduler : TaskScheduler
+        {
+            protected override IEnumerable<Task> GetScheduledTasks()
+            {
+                return Enumerable.Empty<Task>();
+            }
+
+            protected override void QueueTask(Task task)
+            {
+                throw new InvalidOperationException("Use [ForegroundFactAttribute]");
```
Comment (NTaylorMullen) [inline test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundDispatcherTestBase.cs]: typeof is your friend.

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundFactAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+    [XunitTestCaseDiscoverer(nameof(ForegroundFactAttribute), nameof(Microsoft.VisualStudio.LanguageServices.Razor))]
```
Comment (NTaylorMullen) [inline test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundFactAttribute.cs]: This works? I thought you had to provide the actual discoverer.

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Microsoft.VisualStudio.LanguageServices.Razor.Test.csproj [hunk 5]
```diff
@@ -14,6 +14,10 @@
   </ItemGroup>
 
   <ItemGroup>
+    <Reference Include="WindowsBase" />
```
Comment (NTaylorMullen) [inline test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Microsoft.VisualStudio.LanguageServices.Razor.Test.csproj]: What's this for?

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundFactAttribute.cs (hunk 4 above)
Comment (rynowak) [reply test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundFactAttribute.cs]: I don't understand this comment

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Microsoft.VisualStudio.LanguageServices.Razor.Test.csproj (hunk 5 above)
Comment (rynowak) [reply test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Microsoft.VisualStudio.LanguageServices.Razor.Test.csproj]: This is where `System.Windows.Threading.Dispatcher` lives

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundDispatcherTestBase.cs (hunk 3 above)
Comment (rynowak) [reply test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Xunit/ForegroundDispatcherTestBase.cs]: sure

**Pull request merged**